File: src/Grid.cpp
#include "Grid.h"
#include "transform.h"
#include "color.h"

#include <boglfw/renderOpenGL/Shape2D.h>
#include <boglfw/math/math3D.h>

Grid::Grid(int size, int winW, int winH) : size_(size), winW_(winW), winH_(winH) {
}

Grid::~Grid() {
}

void Grid::draw(Viewport*) {
	float squareSz = size_ * tr_.scale;
	int nSqX = ceil(winW_ / squareSz);
	int nSqY = ceil(winH_ / squareSz);
	float offsX = fmodf(tr_.transX, size_);
	if (offsX < 0)
		offsX += size_;
	offsX *= tr_.scale;
	float offsY = fmodf(tr_.transY, size_);
	if (offsY < 0)
		offsY += size_;
	offsY *= tr_.scale;

	// draw horizontal lines
	for (int i=0; i<nSqY; i++) {
		float y = offsY + i * squareSz;
		Shape2D::get()->drawLine({0, y}, {winW_, y}, 0, Colors::GRID);
	}

	// draw vertical lines
	for (int i=0; i<nSqX; i++) {
		float x = offsX + i * squareSz;
		Shape2D::get()->drawLine({x, 0}, {x, winH_}, 0, Colors::GRID);
	}
}

ScreenPoint Grid::gridToScreen(GridPoint p) const {
	int sX = (p.x * size_ + tr_.transX) * tr_.scale;
	int sY = (p.y * size_ + tr_.transY) * tr_.scale;
	return {sX, sY};
}

WorldPoint Grid::gridToWorld(GridPoint p) const {
	return {(float)p.x * size_, (float)p.y * size_};
}

GridPoint Grid::screenToGrid(ScreenPoint p) const {
	return worldToGrid(p.toWorld(tr_));
}

GridPoint Grid::worldToGrid(WorldPoint wp) const {
	float gX = wp.x / size_;
	float gY = wp.y / size_;
	GridPoint ret {(int)gX, (int)gY, 0};
	gX = gX - (int)gX;
	if (abs(gX) > 0.5) {
		ret.x += gX > 0 ? 1 : -1;
		gX = 1.f - abs(gX);
	}
	gY = gY - (int)gY;
	if (abs(gY) > 0.5) {
		ret.y += gY > 0 ? 1 : -1;
		gY = 1.f - abs(gY);
	}
	ret.distance = sqrtf(gX*gX + gY*gY);
	return ret;
}

File: client/src/grid.ts
import { GridPoint, ScreenPoint, Transform, WorldPoint } from "./math/math";
import { fmod } from "./math/functions";

export class Grid {
	private tr_: Transform;

	constructor(private readonly size_: number, private readonly winW_: number, private readonly winH_: number) {}

	setTransform(t: Transform): void {
		this.tr_ = t;
	}
	getTransform(): Transform {
		return this.tr_;
	}
	draw(): void {
		const squareSz = this.size_ * this.tr_.scale;
		const nSqX = Math.ceil(this.winW_ / squareSz);
		const nSqY = Math.ceil(this.winH_ / squareSz);
		let offsX = fmod(this.tr_.transX, this.size_);
		if (offsX < 0) offsX += this.size_;
		offsX *= this.tr_.scale;
		let offsY = fmod(this.tr_.transY, this.size_);
		if (offsY < 0) offsY += this.size_;
		offsY *= this.tr_.scale;

		// draw horizontal lines
		for (let i = 0; i < nSqY; i++) {
			const y = offsY + i * squareSz;
			// FIXME draw
			// Shape2D::get()->drawLine({0, y}, {winW_, y}, 0, Colors::GRID);
		}

		// draw vertical lines
		for (let i = 0; i < nSqX; i++) {
			const x = offsX + i * squareSz;
			// FIXME draw
			// Shape2D::get()->drawLine({x, 0}, {x, winH_}, 0, Colors::GRID);
		}
	}

	cellSize(): number {
		return this.size_;
	}
	gridToScreen(p: GridPoint): ScreenPoint {
		const sX = Math.floor((p.x * this.size_ + this.tr_.transX) * this.tr_.scale);
		const sY = Math.floor((p.y * this.size_ + this.tr_.transY) * this.tr_.scale);
		return new ScreenPoint(sX, sY);
	}

	gridToWorld(p: GridPoint): WorldPoint {
		return new WorldPoint(p.x * this.size_, p.y * this.size_);
	}

	screenToGrid(p: ScreenPoint): GridPoint {
		return this.worldToGrid(p.toWorld(this.tr_));
	}

	worldToGrid(wp: WorldPoint): GridPoint {
		let gX = wp.x / this.size_;
		let gY = wp.y / this.size_;
		const ret = new GridPoint(Math.floor(gX), Math.floor(gY), 0);
		gX = gX - Math.floor(gX);
		if (Math.abs(gX) > 0.5) {
			ret.x += gX > 0 ? 1 : -1;
			gX = 1.0 - Math.abs(gX);
		}
		gY = gY - Math.floor(gY);
		if (Math.abs(gY) > 0.5) {
			ret.y += gY > 0 ? 1 : -1;
			gY = 1.0 - Math.abs(gY);
		}
		ret.distance = Math.sqrt(gX * gX + gY * gY);
		return ret;
	}
}
